src/device_manager.rs

```diff
@@ -0,0 +1,400 @@
+// Copyright © 2019 Intel Corporation. All Rights Reserved.
+// SPDX-License-Identifier: Apache-2.0 and BSD-3-Clause
+
+//! System device management.
+//!
+//! [DeviceManager](struct.DeviceManager.html) responds to manage all devices
+//! of virtual machine, store basic device information like name and
+//! parent bus, register IO resource callback, unregister devices and help
+//! VM IO exit handling.
+
+extern crate vm_allocator;
+
+use self::vm_allocator::SystemAllocator;
+use crate::device::*;
+use std::cmp::{Ord, Ordering, PartialEq, PartialOrd};
+use std::collections::btree_map::BTreeMap;
+use std::collections::HashMap;
+use std::result;
+use std::sync::{Arc, Mutex};
+use vm_memory::{GuestAddress, GuestUsize};
+
+/// Guest physical address and size pair to describe a range.
+#[derive(Eq, Debug, Copy, Clone)]
+pub struct Range(pub GuestAddress, pub GuestUsize);
+
+impl PartialEq for Range {
+    fn eq(&self, other: &Range) -> bool {
+        self.0 == other.0
+    }
+}
+
+impl Ord for Range {
+    fn cmp(&self, other: &Range) -> Ordering {
+        self.0.cmp(&other.0)
+    }
+}
+
+impl PartialOrd for Range {
+    fn partial_cmp(&self, other: &Range) -> Option<Ordering> {
+        self.0.partial_cmp(&other.0)
+    }
+}
+
+/// Error type for `DeviceManager` usage.
+#[derive(Debug)]
+pub enum Error {
+    /// The insertion failed because the new device overlapped with an old device.
+    Overlap,
+    /// The insertion failed because the resource is not enough.
+    Oversize,
+    /// PIO address is none.
+    NonePIOAddress,
+    /// The insertion failed because device already exists.
+    Exist,
+    /// The removing fails because the device doesn't exist.
+    NonExist,
+    /// IRQ allocated failed.
+    AllocateIrq,
+}
+
+/// Simplify the `Result` type.
+pub type Result<T> = result::Result<T, Error>;
+
+/// System device manager serving for all devices management and VM exit handling.
+pub struct DeviceManager<'a> {
+    /// System allocator reference.
+    resource: &'a mut SystemAllocator,
+    /// Devices information mapped by name.
+    devices: HashMap<String, DeviceDescriptor>,
```

Comment:
@serban300 For the vmm usage scenario, I think it is reasonable to assume different names for different devices to record the basic device information, which might be used to dump device tree or something like this.

Reply:
That's not the case, one example is one device type can have multiple instances, e.g. if you don't include the instance_id as part of name, then those instances will have the same name. With the current Device trait, the device itself may even don't know the instance_id so can't include that when the name() gets called.

I prefer to define a lighter, unique device ID for each device that we can allocate the ID in a center place to guarantee the ID is unique across a VM instance and the program refers the device using this unique ID .

Reply:
@serban300 Right, previously I though passing a name to e.g. `VirtioBlk::new("virtio-blk-1") ` for each device type to make different instances different names. According to your and @chao-p 's suggestion, I agree that we define a light struct might be like this,
```
struct device_info {
    device_type: String, // Actually a name of type e.g. virtio-blk
    instance_id: u32,
}
```
We allocate a global instance_id to every device instance from number 1, and put this struct into `DeviceDescriptor`, replacing the first member currently. What do you think? :)

Reply:
Using a structure would be perfect for us. It would be consistent with what we already have in Firecracker.

I think the device_type can be something lighter than a string. For example in Firecracker we use an enum:

```
/// Types of devices that can get attached to this platform.
#[derive(Clone, Debug, PartialEq, Eq, Hash)]
pub enum DeviceType {
    Virtio(u32),
    #[cfg(target_arch = "aarch64")]
    Serial,
    #[cfg(target_arch = "aarch64")]
    RTC,
}
```
Where the u32 associates to the Virtio can be:
```
/// Types taken from linux/virtio_ids.h.
pub const TYPE_NET: u32 = 1;
pub const TYPE_BLOCK: u32 = 2;
pub const TYPE_VSOCK: u32 = 19;
```

In our case the instance_id is a String provided by the user when the vm is configured. That is unique per device type. Having a key composed of device_type and this instance_id makes it easy for us to retrieve the device when the user sends a request for changing some parameters (changing the file path for a block device for example). So do you think the instance_id could be a String in your case also ? Or do you need it to be a global counter ?

Reply:
@serban300 Thanks for the example for device_type.  That is a good way to be light!
Umm, till now, the issue seems different with beginning :) It seems we feel a name string (e.g. DummyDevice::new(name)) can not ensure a unique per instance, so we extend it to a structure.
If we have structure {DeviceType , instance_id_string}, I guess we go back to the problem again? :-p

Reply:
Yes, true ...

Again an enum would offer the flexibility of chosing what you need depending on the usecase, but it would be heavier. It might also look hacky. I don't know. Do you think it would make sense ?

```
pub enum DeviceInstanceId {
    GlobalCounter(u32),
    Name(String),
}
```

Reply:
@serban300 If the GlobalCounter(idx) is offered by users, or allocated by system?
I guess, we have an allocator, specifying idx to each new instance. But this needs it to be a struct... Something like,
```
pub struct InstanceId {
    pub last_id: u32,
}
impl InstanceId {
    /// Called by system
    pub fn new(start: u32) -> Self {
        InstanceId {last_id: start},
    }
    /// Get a u32 for each instance info {device_type: enum DeviceType, instance_id: u32}
    pub fn allocate() -> u32 {
        self.last_id = self.last_id + 1;
        self.last_id - 1
    }
}

Reply:
It should be allocated by the system in order to guarantee its uniqueness. An allocator struct would work.

Reply:
How do you guarantee the uniqueness? If you use the `rand` crate in Rust it pulls a lot of dependencies. If we have to use rand, I would rather have the unique ID specified by the user of the crate so we don't pull all those dependencies in this library.

Reply:
@andreeaflorescu Why do we need `rand` crate? I think an allocator struct above can help us allocate a unique number to each instance. Do I miss anything?

Reply:
@liujing2 I totally missed the example code from your previous comment. It looks like we don't need rand.

Some comments:
- we need to make sure the id doesn't overflow
- `allocate` might be called from different threads

Reply:
A few comments/questions:

* The ID allocation API would belong to the `vm-allocator`. As @andreeaflorescu pointed out, it should handle concurrency. Maybe implement this API as wrapper on top of the `Atomic` APIs?

* Are we agreeing on the fact that the key to the `devices` hash should be a structure, and not a string:

```Rust
pub struct DeviceId {
     /// An allocated ID from vm-allocator.
     /// Should we have a specific type for it?
     id: u32

     /// Device friendly name
     name: String
}
```

* Or should we instead have `vm-allocator` provide a naming API that would generate unique strings (given a prefix), and use that as a unique name for the device?

Reply:
@sameo I am a bit behind we the design of vm-device and vm-allocator. Do we want vm-device to take a dependency on vm-allocator?

Reply:
> Some comments:
> 
> * we need to make sure the id doesn't overflow
> * `allocate` might be called from different threads

@andreeaflorescu Yep, a u32 type for id can be enough. 
@sameo I agreed to make the id allocator in vm-allocator crate.

Reply:
> A few comments/questions:
> 
> * The ID allocation API would belong to the `vm-allocator`. As @andreeaflorescu pointed out, it should handle concurrency. Maybe implement this API as wrapper on top of the `Atomic` APIs?
> * Are we agreeing on the fact that the key to the `devices` hash should be a structure, and not a string:
> 
> ```rust
> pub struct DeviceId {
>      /// An allocated ID from vm-allocator.
>      /// Should we have a specific type for it?
>      id: u32
> 
>      /// Device friendly name
>      name: String
> }
> ```

According to above discussion with @serban300 's , they might prefer using enum instead of String as the name.

> * Or should we instead have `vm-allocator` provide a naming API that would generate unique strings (given a prefix), and use that as a unique name for the device?

I prefer to this idea. An API for allocating instance id (u32).

Reply:
If vm-device doesn't already have a dependency on vm-allocator I wouldn't add one just for this reason. We can have a DeviceId that needs to be passed when initializing a new device and let the caller decide how to initialize it. One might choose to use vm-allocator for that, another one can choose to set the device id using a different mechanism.

Reply:
> According to above discussion with @serban300 's , they might prefer using enum instead of String as the name.

That would give names like e.g. `virtio-net-afb3467`, once the enum is stringified. Enums can be very generic as opposed to something more grepable based on a friendly, custom name provided by the device implementation as e.g. a prefix.

Reply:
> @sameo I am a bit behind we the design of vm-device and vm-allocator. Do we want vm-device to take a dependency on vm-allocator?

It  does have a dependency on it already, this PR comes with a vm-allocator implementation for allocating and tracking pin based irqs and IO address spaces.

Reply:
@sameo, do you plan to submit vm-allocator crate adding this new API? Or I can update assuming there is one.

Reply:
@liujing2 Please continue updating the vm-allocator code that's part of this PR, we will base the crate on it.